### ISSUE TITLE
Mark-as-reviewed with per-line granularity

### ISSUE THREAD

**Issue by maddog-tri** [Reviewable/Reviewable#414]
Currently, mark-as-reviewed applies to entire files.  It would be super-useful to allow individual lines to be marked as reviewed.  This is handy for a "multiple pass" style of reviewing, where one quickly scans the PR and checks off changes that look ok ("low hanging fruit"), and then one returns for another pass to see what is left over (the not low-hanging fruit).  There is currently no way to indicate "this change does not need further review by me".

It'd be nice to be able to select/mark entire blocks of code at one time --- but just being able to perform a single-click mark on individual lines would get the job done.

**Comment by pkaminski** [comment]
While I see the value in having finer granularity check-offs, I think the UI would become overwhelming and the technical challenge would be significant as well.  We'd need some affordance for marking lines (or sections) as reviewed, a way to indicate each line's state, and probably modifications to all other places where reviewed files are mentioned.  On the technical side, instead of just diffing against the last-reviewed revision, we'd now need to track things at the line level and ensure that previously reviewed lines that haven't changed aren't shown as different in the new revision either.

Ultimately, I think such a feature would be overwhelmingly complex, and many users already fine the current UI too complex to understand.  Sorry.

**Comment by pkaminski** [comment]
So it's 5 years later and I'm thinking this might be both feasible and worth doing after all.  My idea is that you'd be able to select a range of lines in the diff and some buttons would pop up attached to it:  copy code, create code block comment, and mark reviewed.  Marking a range as reviewed would mark those lines (on both sides of the diff) as reviewed for the specific pair of commits being diffed.  If the same pair of commits is diffed in the future those lines would be elided from the diff and the break blurb would include "NN lines previously reviewed" to expand them anyway, similar to how, e.g., base changes or whitespace changes are elided.  This wouldn't affect any other revision pairs or the reviewed status of a file, and wouldn't be visible to other users.

How does that sound?  (Also ping @lakshayg from #728.)

**Comment by maddog-tri** [comment]
Finally!

I must admit that I haven't used Reviewable in just about three years (i.e., since I left TRI, _forever ago_), but, yeah, your proposal sounds good.  My primary goal for this feature was to be able to incrementally record my personal progress while reviewing a file (often in a non-linear, random-access traversal), to ensure that I did in fact inspect everything before signing off on the file/changeset.  This should do the trick.  Having the per-line marks be not visible to other users nor affect the overall review status, as you have proposed, are the right way to do it, IMO.

**Comment by lakshayg** [comment]
This sounds like a good idea!

**Comment by jwnimmer-tri** [comment]
For those who haven't seen it yet, #892 is somewhat related here (in terms of workflow).  When I'm reviewing a large change, I mark things like "??? revisit this function" during review.  Marking lines of a diff to elide in the review (as proposed here) might also be nice, but for my part the "notes to myself" has been good enough so far.

**Comment by liorgold2** [comment]
This feature would be extremely helpful!
1. When reviewing a file with multiple changes, it often makes sense to review some chunks first, make some comments, and continue the review only after the author fixes the comments.
2. It helps to be able to hide the things you finished reviewing so that only the remaining parts show. Very similarly to the effect of marking whole files as reviewed, except that here you don't need to complete the review of a file (which is annoying in some PRs).

I think that even a non-perfect version of this feature can be helpful, and this can be improved over time if needed. Since this is indeed an advanced feature it makes sense to hide it to make sure it doesn't bother people who're just starting to use Reviewable.

**Comment by pkaminski** [comment]
Yup, though note that the feature will only address your second use case, not this one:

> 1. When reviewing a file with multiple changes, it often makes sense to review some chunks first, make some comments, and continue the review only after the author fixes the comments.

Line-specific marks will be tightly bound to a revision, so if the author pushes some fixes the line marks will not carry across to the new revision.  (Cross-revision line mapping, like we do for discussions, is too imprecise and much too likely to hide unreviewed lines if applied to line marks.)  The partial review use case is covered by our [deferral feature](https://docs.reviewable.io/reviews.html#deferring-a-review).
